### ISSUE TITLE
Problems with the TestConverter and ideas how to solve them

### ISSUE THREAD

**Issue by moonglum** [CodePadawans/ataru#62]
Hey Padawans,

After my mysterious meeting today with @railsbros-dirk we would like to suggest some changes to the way that the TestConverter works. We currently see two problems with it:
1. A bug, which is in the concept and not in the implementation.
2. It is too complicated and could be done in a more straight forward way.
## The Bug

Given you have a code sample with more than one "comment arrow", only one of the "comment arrows" will be found and returned to the user. The converter only returns one 'assert_equal' statement in the form of the `[:ok, 'assert_equal ...']` statement. This leads also to the problem I discussed with @madziaf that you somehow have to attach the assertion to the end of the statement. Take the following Ruby Code:

``` ruby
a = 1

a #=> 1

a = 2
```

This is absolutely valid and should return one assertion that succeeds. What Ataru would do right now is to convert it to the following snippet:

``` ruby
a = 1

a #=> 1

a = 2
assert_equal 1, a
```

Which will result in an error, because a is 2 now. I have no idea on how to solve this in the current concept. So I came up with a new concept.
## The new concept

The CodeSample class gets an additional attribute 'prepared_code'. The TestConverter gets a code_sample object as an argument to the convert method as it does right now. This method does not return anything, but it will set the 'prepared_code' attribute. This 'prepared_code' attribute now contains the code that is ready to be evaluated, including all the assert_equal statements. When the attribute `code` is the following:

``` ruby
a = 1
a #=> 1

a = 2
a #=> 2
```

It will set the `prepared_code` to the following value:

``` ruby
a = 1
assert_equal 1, a

a = 2
assert_equal 2, a
```

It **replaces** all arrow lines with the equivalent `assert_equal` line. In `lib/ataru/application.rb` you can now replace the entire `code_samples.map ...` thing with something that just iterates over all the samples and calls `TestConvert.convert` on each of the CodeSample objects. Clear so far?

Now how does the convert function work? There's a nice method in Ruby called [`gsub`](http://ruby-doc.org/core-2.1.2/String.html#method-i-gsub) which will help us do the job. It replaces all instances of something with something else. The magical thing about it is that it takes an RegEx as the thing that you want to replace. Let me give you an example!

``` ruby
"a b aa bbbb aaa".gsub(/a+/, 'PADAWANS!!!') # => "PADAWANS!!! b PADAWANS!!! bbbb PADAWANS!!!"
```

You see that it replaces **all occurrences** of the pattern with the second thing. We can now use that on our example from above to replace all occurrences of the arrow with the according `assert_equal` statement. At this point I want you to scream at me "LUCAS WHAT ARE YOU TALKING ABOUT" if anything is not clear.

Now, let's get a little fancy with our example! Because one thing we need is that we need to know the exact match we have found and use it in our replacement String. So given we have the tasks of converting all occurrences of `a+b` where a and b can be anything with `b+a` – how would we do that?

``` ruby
a = "c+a + a c d+e f+c"

a.gsub(/(.)\+(.)/, '\2+\1') # => "a+c + a c e+d c+f"
```

At this point you're probably saying: WHAT THE HELL? And yes, that is quite appropriate. In the replacement String you get some special powers (only if you are using single quotes though – just to save you some trouble) – in the form of numbers prefixed with a slash. Those are the **matches** from your regular expression. The matches are the things that you put parentheses around. So `\1` is the first thing that is wrapped in parentheses, and `\2` is the second thing that is wrapped in parentheses. In the String you now just put the second match then a plus and then the first match. And voila: You have swapped those two :smile: Does that make any sense?

Now, my challenge to you is to write a gsub expression that would convert the following Ruby code:

``` ruby
a = 1
a #=> 1

a = 2
a #=> 2
```

into this Ruby code:

``` ruby
a = 1
assert_equal 1, a

a = 2
assert_equal 2, a
```

Do you feel up to the challenge? If so, please try it out and we will then continue from there with our `convert` function, which is then almost done! If not, please ask away!!

**Comment by code-later** [comment]
![giphy copy](https://cloud.githubusercontent.com/assets/1948/4317946/116115e4-3f19-11e4-9fab-6a653086f685.gif)

**Comment by nanoparsec** [comment]
Challenge accepted. How about this solution?

``` ruby
foo = "a = 1 a #=> 1"
bar = "a = 1 a # => 1"

pup = foo.gsub(/#=>|# =>/, 'assert_equal')
poop = bar.gsub(/#=>|# =>/, 'assert_equal')

pup.gsub(/(.) assert_equal (.)/, 'assert_equal \1 ,\2')
```

**Comment by moonglum** [comment]
Very good :+1: :dancers:  So this would be a two step process, right? So when you start with this:

``` ruby
a = 1
a #=> 1
```

It will first convert it to this:

``` ruby
a = 1
a assert_equal 1
```

And then in a second step:

``` ruby
a = 1
assert_equal a, 1
```

Correct?

PS: I adjusted the code above a little bit after @madziaf's feedback in HipChat.

**Comment by nanoparsec** [comment]
Jup.

**Comment by moonglum** [comment]
Great work! 

![giphy](https://cloud.githubusercontent.com/assets/95788/4321619/be342d9c-3f41-11e4-9360-430e184ecf3b.gif)

Do you think that you can also manage this in one step, with one `gsub` call? :smile: I think you two can do it!

**Comment by moonglum** [comment]
If you want a tiny hint:

There is an easier way to say this: `#=>|# =>` You can use the questionmark I showed you in the screencast to say "there could be either one or zero of the character I just wrote". If you have that solution, you can take this and replace the `assert_equal` in your second RegExp :+1:

**Comment by nanoparsec** [comment]
p foo.gsub(/(.) # ?=> (.)/, 'assert_equal \2, \1')

**Comment by moonglum** [comment]
Yeeeeees!!!! :smile: :smile:

**Comment by moonglum** [comment]
Ok, now a last tweak, and then it is perfect :dancers: :dancers: 

It should also be able to do the following:

``` ruby
longer_name = 12
longer_name #=> 12
```

to

``` ruby
longer_name = 12
assert_equal 12, longer_name
```

**Comment by plexus** [comment]
will this code work?

``` ruby
x = " #=>
" #=> " #=>\n"
```

**Comment by moonglum** [comment]
@plexus Yes, that would lead to problems. But the only solution to that would be using a Ruby parser to find all comments etc. – this can still be done later. This will handle 99% of the cases.

**Comment by plexus** [comment]
yeah I know :) I'm sensitive to seeing formal languages manipulated with regular expressions, but in this stage I agree it's a valid trade-off.

**Comment by moonglum** [comment]
@plexus I know, we talked about that last year at eurucamp and I totally agree :wink:

**Comment by nanoparsec** [comment]
p bar.gsub(/(\S+) # ?=> (.+)/, 'assert_equal \2, \1')

**Comment by moonglum** [comment]
Yay!! That's great! Amazing work, you two!!

**Comment by nanoparsec** [comment]
p foo.gsub(/(\S+) # ?=> (\S+)/, 'assert_equal \2, \1')

We think this would be better?

**Comment by moonglum** [comment]
Yep :+1:

**Comment by madziaf** [comment]
solved with #64
